### PR TITLE
[5.4] Add new JSON loader for translations

### PR DESCRIPTION
**Pull request by themsaid**
This PR introduces a new form of handling localization that's based on .json files instead of current PHP array files, the new format is similar to the popular gettext format except that it loads from json instead of po files.

```php
__('Product ":product" was saved successfully.', [$product->name]);
```

This line will load the `resources/lang/en.json` and look for a `Product ":product" was saved successfully.` key inside it, if key wasn't found it just returns the given string.

The replacement of parameters happens even if the key wasn't found so that you don't have to have translation files for your main language, using the keys only would work.

If no translation was found in the JSON files, Laravel will start looking for translation keys in the old PHP array files located in `resources/lang/en`, in case a match was found it's returned to the user. That way the method is fully backward compatible.

The old translation methods will work as they used to `trans()` & `trans_choice()`, these PR doesn't look into deprecating these methods, it just introduces an alternative approach.

### CODE REVIEW

File: src/Illuminate/Translation/FileLoader.php [hunk 1]
```diff
@@ -114,6 +118,22 @@ protected function loadPath($path, $locale, $group)
...
+     * Load a locale from a given JSON file.
+     *
+     * @param  string  $path
+     * @param  string  $locale
+     * @return array
+     */
+    protected function loadJson($path, $locale)
+    {
+        if ($this->files->exists($full = "{$path}/{$locale}.json")) {
+            return (array) json_decode($this->files->get($full));
```
Comment (CristianLlanos) [inline src/Illuminate/Translation/FileLoader.php]: Can we use `json_decode($this->files->get($full), true)` instead? It'll give us an associative array [[1](http://php.net/json_decode)] :)

Comment (themsaid) [inline src/Illuminate/Translation/FileLoader.php]: Thank you, changed :)

**Comment by tillkruss** [comment]
@themsaid: Maybe the `__()` helper should also use the `array` type hint, just like `getFromJson()` does?

**Comment by dionysiosarvanitis** [comment]
Why not using something more fluent over one-letter functions? __() seems kind of noisy or at least use t().
Issue is related to https://github.com/laravel/internals/issues/252.

**Comment by kduma** [comment]
I think this PR provides unnecessary mess and redundancy with old `trans` function.

**Comment by themsaid** [comment]
@kduma how?

**Comment by taylorotwell** [comment]
Actually “__” was the name of the function in Laravel 1.0 - 3.0. #trivia :)

**Comment by kduma** [comment]
@themsaid now there are two ways of storing translations, the current way in PHP files (`trans` function) and new way, in json files ('__' function).

**Comment by kduma** [comment]
@taylorotwell in my first PHP projects, I've also used `__` functions for translations. (It was in times of free Translation API from google)

**Comment by kkomelin** [comment]
Glad to see the Laravel translation system improving. Thanks.

**Comment by sisve** [comment]
1. The title implies that there is a new json loader, but there isn't. Why wasn't this implemented as a new JsonFileLoader (implementing LoaderInterface)?
2. Why is __ introduced as a new translation function that's hardcoded to work against a specific storage format?

**Comment by taylorotwell** [comment]
1. What's the difference?

2. It doesn't really matter to me what the function name is. __ is what the translation function was called previously in Laravel. I haven't though through the implementations of having `trans` check the JSON translations first and then fall back to the grouped array files but it may be a possibility.

**Comment by sisve** [comment]
1. Structure, mostly. There's already a ArrayLoader and a FileLoader. I doubt the file-loader will be extended to support all future file formats; I consider it more appropriate with a JsonLoader in this case.
2. I do not have a problem with a __ function for translation, the problem, as I see it, is that it is hardcoded to a specific type of translation thingie (it calls `$translator->getJson()`) which ends up with a `json_decode($this->files->get($full), true)` call that basically circumvents the existing LoaderInterface structure.

This, combined with the makeJsonReplacements method, looks odd. Doesn't trans() already support this parameter substitution? Why wasn't that code reused?

**Comment by kkomelin** [comment]
@sisve Although your questions/doubts might be reasonable, it would be more constructive if you'd propose your improvements instead of questioning solutions of others. 
Opensource means collaboration. If we think something is not perfect or we'd implement it in a better way, we propose our improvements through issues or PRs. I encourage you to do the same. Thanks.

**Comment by taylorotwell** [comment]
makeJsonReplacements was removed.

**Comment by taylorotwell** [comment]
@sisve we welcome your improvements to the loader setup via PR.

**Comment by cjmaxik** [comment]
So, what about piped plural strings?
```{0}Nobody playing|:count player|:count players```

**Pull request merged**